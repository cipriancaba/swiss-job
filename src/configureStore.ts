import { applyMiddleware, createStore } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import { persistCombineReducers, PersistedState, persistStore } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import createSagaMiddleware from 'redux-saga'
import { dictionaries, DictionariesState } from './reducers/dictionaries.reducer'
import { originalDataset, OriginalDatasetState } from './reducers/originalDataset.reducer'
import { rootSagas } from './sagas'

export interface RootState extends PersistedState {
  originalDataset: OriginalDatasetState
  dictionaries: DictionariesState
}

const initialState: Partial<RootState> = {}
const sagaMiddleware = createSagaMiddleware()

const middlewares = [sagaMiddleware]
const middleware = applyMiddleware(...middlewares)

const config = {
  key: 'root',
  storage,
  version: 1,
  whitelist: ['originalDataset', 'dictionaries'],
}

const reducersWithPersist = persistCombineReducers(config, { originalDataset, dictionaries })
const store = createStore(reducersWithPersist, initialState, composeWithDevTools(middleware))

const persistor = persistStore(store)
persistor.purge()

sagaMiddleware.run(rootSagas)

export { store, persistor }
