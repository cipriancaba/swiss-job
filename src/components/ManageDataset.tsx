import * as React from 'react'
import { connect, DispatchProp } from 'react-redux'
import { datasetActions } from '../actions/dataset.actions'
import { RootState } from '../configureStore'
import { Product, ValidatedProduct } from '../types'
import { EditableTable } from './EditableTable'

type ManageDatasetProps = ReturnType<typeof mapStateToProps> & DispatchProp

class ManageDatasetComponent extends React.PureComponent<ManageDatasetProps> {
  private getValidatedProduct = (
    property: keyof Product,
    newValue: string,
    product: ValidatedProduct<Product>,
    products: Array<ValidatedProduct<Product>>
  ) => {
    const newProduct = {
      ...product,
      [property]: {
        value: newValue,
        isValid: !!newValue,
        validationError: newValue === '' ? 'Please enter a value' : undefined,
      },
    }
    return newProduct
  }

  private createNewProduct = () => ({
    product: '',
    color: '',
    price: '',
  })

  private onUpdateData = (products: Array<ValidatedProduct<Product>>) => {
    this.props.dispatch(
      datasetActions.updateDictionary(
        products.map(p => ({ product: p.product.value, color: p.color.value, price: p.price.value }))
      )
    )
  }

  public render = () => {
    return (
      <div className="ManageDataset">
        <h1>MANAGE DATASET</h1>
        <div>
          <EditableTable<Product>
            products={this.props.products}
            getValidatedProduct={this.getValidatedProduct}
            createNewProduct={this.createNewProduct}
            headers={['Product', 'Color', 'Price']}
            onUpdateData={this.onUpdateData}
          />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state: RootState) => {
  return { products: state.dataset }
}

export const ManageDataset = connect(mapStateToProps)(ManageDatasetComponent)
