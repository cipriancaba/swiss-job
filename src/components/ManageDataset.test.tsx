import { configure, shallow } from 'enzyme'
import ReactSixteenAdapter from 'enzyme-adapter-react-16'
import * as React from 'react'
import { Provider } from 'react-redux'
import { store } from '../configureStore'
import { ManageDataset } from './ManageDataset'

configure({ adapter: new ReactSixteenAdapter() })

describe('ManageDataset', () => {
  it('renders ManageDataset', () => {
    const manageDatasetPage = shallow(
      <Provider store={store}>
        <ManageDataset />
      </Provider>,
      { context: { store } }
    ).dive()
    expect(manageDatasetPage).toMatchSnapshot()
  })
})
