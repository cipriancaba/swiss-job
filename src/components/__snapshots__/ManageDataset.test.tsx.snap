// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDataset renders ManageDataset 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ManageDatasetComponent) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "products": Array [
        Object {
          "color": "Anthracite",
          "price": "CHF 769",
          "product": "Apple iPhone 6s",
        },
        Object {
          "color": "Midnight Black",
          "price": "CHF 569",
          "product": "Samsung Galaxy S8",
        },
        Object {
          "color": "Mystic Silver",
          "price": "CHF 272",
          "product": "Huawei P9",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "products": Array [
          Object {
            "color": "Anthracite",
            "price": "CHF 769",
            "product": "Apple iPhone 6s",
          },
          Object {
            "color": "Midnight Black",
            "price": "CHF 569",
            "product": "Samsung Galaxy S8",
          },
          Object {
            "color": "Mystic Silver",
            "price": "CHF 272",
            "product": "Huawei P9",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
    },
  },
}
`;
